File: pkg/parser/parser.go
// Copyright (c) 2020, The Facette Authors
//
// Licensed under the terms of the BSD 3-Clause License; a copy of the license
// is available at: https://opensource.org/licenses/BSD-3-Clause

// Package parser provides expressions lexical parsing.
package parser

import (
	"bufio"
	"bytes"
	"fmt"
	"io"
)

const eof = rune(0)

// Parser is a lexical parser.
type Parser struct {
	scanner   io.RuneScanner
	pos, last Position
	peeked    *Token
}

// New creates a new lexical parser instance.
func New(r io.Reader) *Parser {
	return &Parser{
		scanner: bufio.NewReader(r),
		pos:     Position{Line: 1, Char: 1},
	}
}

// Expect returns the next token ensuring for its type.
func (p *Parser) Expect(typ Type) (*Token, error) {
	tok := p.Next()
	if tok.Type != typ {
		return tok, fmt.Errorf("expected %s but got %s at %s", typ, tok.Type, tok.Pos)
	}

	return tok, nil
}

// Next returns the next token.
func (p *Parser) Next() *Token {
	var tok *Token

	if p.peeked != nil {
		tok = p.peeked
		p.peeked = nil
	} else {
		tok = p.scan()
	}

	return tok
}

// Peek returns the next token without consuming it.
func (p *Parser) Peek() *Token {
	if p.peeked == nil {
		tok := p.scan()
		p.peeked = tok
	}

	return p.peeked
}

// PeekRune returns the next rune without consuming it.
// FIXME: find a way to ditch this as it only serves for function ident
//        detection in series expression.
func (p *Parser) PeekRune() rune {
	r, _ := p.read()
	p.unread()

	return r
}

func (p *Parser) read() (rune, Position) {
	ch, _, err := p.scanner.ReadRune()
	if err != nil {
		return eof, p.pos
	}

	p.last = p.pos

	if ch == '\n' {
		p.pos.Line++
		p.pos.Char = 1
	} else {
		p.pos.Char++
	}

	return ch, p.last
}

func (p *Parser) run(set []byte) []byte {
	b := bytes.NewBuffer(nil)

	for {
		ch, _ := p.read()
		if !bytes.ContainsRune(set, ch) {
			break
		}

		b.WriteRune(ch)
	}

	p.unread()

	return b.Bytes()
}

func (p *Parser) scan() *Token {
	c, pos := p.read()
	for isSpace(c) {
		c, pos = p.read()
	}

	switch {
	case c == eof:
		return newToken(EOF, pos, "")

	case c == '-' || isDigit(c):
		p.unread()
		return p.scanNumber()

	case isIdentChar(c):
		p.unread()
		return p.scanIdent()

	case c == '"' || c == '\'':
		p.unread()
		return p.scanString()

	case c == '=':
		next, _ := p.read()
		if next == '~' {
			return newToken(EQREGEXP, pos, "")
		}

		p.unread()

		return newToken(EQ, pos, "")

	case c == '!':
		next, _ := p.read()
		if next == '=' {
			return newToken(NEQ, pos, "")
		} else if next == '~' {
			return newToken(NEQREGEXP, pos, "")
		}

		p.unread()

	case c == '{':
		return newToken(LBRACE, pos, "")

	case c == '}':
		return newToken(RBRACE, pos, "")

	case c == '(':
		return newToken(LPAREN, pos, "")

	case c == ')':
		return newToken(RPAREN, pos, "")

	case c == ',':
		return newToken(COMMA, pos, "")
	}

	return newToken(INVALID, pos, string(c))
}

func (p *Parser) scanIdent() *Token {
	start := p.pos

	buf := bytes.NewBuffer(nil)

loop:
	for {
		ch, _ := p.read()
		switch {
		case ch == eof:
			break loop

		case isIdentChar(ch):
			buf.WriteRune(ch)

		default:
			p.unread()
			break loop
		}
	}

	return newToken(IDENT, start, buf.String())
}

func (p *Parser) scanNumber() *Token {
	digits := []byte("0123456789")
	start := p.pos

	buf := bytes.NewBuffer(nil)

	next, _ := p.read()
	if next == '-' {
		buf.WriteRune(next)
	} else {
		p.unread()
	}

	buf.Write(p.run(digits))

	next, _ = p.read()
	if next == '.' {
		buf.WriteRune(next)
		buf.Write(p.run(digits))
	} else {
		p.unread()
	}

	next, _ = p.read()
	if bytes.ContainsRune([]byte("eE"), next) {
		buf.WriteRune(next)

		next, _ = p.read()
		if bytes.ContainsRune([]byte("+-"), next) {
			buf.WriteRune(next)
		} else {
			p.unread()
		}

		buf.Write(p.run(digits))
	} else {
		p.unread()
	}

	return newToken(NUMBER, start, buf.String())
}

func (p *Parser) scanString() *Token {
	quote, start := p.read()

	buf := bytes.NewBuffer(nil)

	for {
		ch, pos := p.read()

		switch ch {
		case quote:
			return newToken(STRING, start, buf.String())

		case eof:
			return newToken(EOF, pos, buf.String())

		case '\n':
			return newToken(NEWLINE, pos, buf.String())

		case '\\':
			next, _ := p.read()
			switch next {
			case '\\', '"', '\'':
				buf.WriteRune(next)

			default:
				return newToken(BADESCAPE, pos, string(ch+next))
			}

		default:
			buf.WriteRune(ch)
		}
	}
}

func (p *Parser) unread() {
	err := p.scanner.UnreadRune()
	if err == nil {
		p.pos = p.last
	}
}

func isAlpha(c rune) bool {
	return c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z'
}

func isDigit(c rune) bool {
	return c >= '0' && c <= '9'
}

func isIdentChar(c rune) bool {
	return isDigit(c) || isAlpha(c) || c == '_'
}

func isSpace(c rune) bool {
	return c == ' ' || c == '\t' || c == '\n'
}

File: ui/src/helpers/parser.ts
/**
 * Copyright (c) 2020, The Facette Authors
 *
 * Licensed under the terms of the BSD 3-Clause License; a copy of the license
 * is available at: https://opensource.org/licenses/BSD-3-Clause
 */

const alpha = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";

const digits = "0123456789";

export function formatPosition(pos: Position): string {
    return `${pos.line}:${pos.char}`;
}

function isDigit(c: string): boolean {
    return digits.includes(c);
}

function isIdentChar(c: string): boolean {
    return (alpha + digits + "_").includes(c);
}

function isSpace(c: string): boolean {
    return " \t\n".includes(c);
}

export interface Token {
    type: TokenType;
    pos: Position;
    text: string;
}

export enum TokenType {
    INVALID = "invalid",
    EOF = "end of input",

    IDENT = "ident", // ident
    NUMBER = "number", // 123.45
    STRING = "string", // "abc" or 'abc'
    NEWLINE = "new line", // \n
    BADESCAPE = "bad escape", // \b

    EQ = "equal", // =
    NEQ = "not equal", // !=
    EQREGEXP = "equal pattern", // =~
    NEQREGEXP = "not equal pattern", // !~

    LBRACE = "left brace", // {
    RBRACE = "right brace", // }
    LPAREN = "left parenthesis", // (
    RPAREN = "right parenthesis", // )
    COMMA = "comma", // ,
}

export interface Position {
    line: number;
    char: number;
}

export class Parser {
    private last: Position | null = null;

    private lastCh: string | null = null;

    private peeked: Token | null = null;

    private pos: Position = {line: 1, char: 1};

    private text = "";

    public constructor(text: string) {
        this.text = text;
    }

    public expect(type: string): Token {
        const tok = this.next();
        if (tok.type !== type) {
            throw Error(`expected ${type} but got ${tok.type} at ${formatPosition(tok.pos)}`);
        }

        return tok;
    }

    public next(): Token {
        let tok: Token;

        if (this.peeked !== null) {
            tok = this.peeked;
            this.peeked = null;
        } else {
            tok = this.scan();
        }

        return tok;
    }

    public peek(): Token {
        if (this.peeked === null) {
            this.peeked = this.scan();
        }

        return this.peeked;
    }

    private read(): string {
        const c: string = this.text.slice(0, 1);
        this.text = this.text.slice(1);

        this.last = Object.assign({}, this.pos);
        this.lastCh = c;

        if (c === "\n") {
            this.pos.line++;
            this.pos.char = 1;
        } else {
            this.pos.char++;
        }

        return c;
    }

    private run(set: string): string {
        let s = "";

        for (;;) {
            const c = this.read();
            if (!set.includes(c)) {
                break;
            }

            s += c;
        }

        return s;
    }

    private scan(): Token {
        let c = this.read();
        while (isSpace(c)) {
            c = this.read();
        }

        if (c === "") {
            return this.tokenAtPos(TokenType.EOF);
        } else if (c === "-" || isDigit(c)) {
            this.unread();
            return this.scanNumber();
        } else if (isIdentChar(c)) {
            this.unread();
            return this.scanIdent();
        } else if (c === '"' || c === "'") {
            this.unread();
            return this.scanString();
        } else if (c === "=") {
            const next = this.read();
            if (next === "~") {
                return this.tokenAtPos(TokenType.EQREGEXP);
            }

            this.unread();

            return this.tokenAtPos(TokenType.EQ);
        } else if (c === "!") {
            const next = this.read();
            if (next === "=") {
                return this.tokenAtPos(TokenType.NEQ);
            } else if (next === "~") {
                return this.tokenAtPos(TokenType.NEQREGEXP);
            }

            this.unread();
        } else if (c === "{") {
            return this.tokenAtPos(TokenType.LBRACE);
        } else if (c === "}") {
            return this.tokenAtPos(TokenType.RBRACE);
        } else if (c === "(") {
            return this.tokenAtPos(TokenType.LPAREN);
        } else if (c === ")") {
            return this.tokenAtPos(TokenType.RPAREN);
        } else if (c === ",") {
            return this.tokenAtPos(TokenType.COMMA);
        }

        return this.tokenAtPos(TokenType.INVALID, c);
    }

    private scanIdent(): Token {
        let s = "";

        for (;;) {
            const c = this.read();
            if (c === "") {
                break;
            } else if (isIdentChar(c)) {
                s += c;
            } else {
                this.unread();
                break;
            }
        }

        return this.tokenAtPos(TokenType.IDENT, s);
    }

    private scanNumber(): Token {
        let s = "";

        let next = this.read();
        if (next === "-") {
            s += next;
        } else {
            this.unread();
        }

        s += this.run(digits);

        next = this.read();
        if (next === ".") {
            s += next + this.run(digits);
        } else {
            this.unread();
        }

        next = this.read();
        if ("eE".includes(next)) {
            s += next;

            next = this.read();
            if ("+-".includes(next)) {
                s += next;
            } else {
                this.unread();
            }

            s += this.run(digits);
        } else {
            this.unread();
        }

        return this.tokenAtPos(TokenType.NUMBER, s);
    }

    private scanString(): Token {
        const quote = this.read();

        let s = "";

        for (;;) {
            const c = this.read();

            switch (c) {
                case quote:
                    return this.tokenAtPos(TokenType.STRING, s);

                case "":
                    return this.tokenAtPos(TokenType.EOF, s);

                case "\n":
                    return this.tokenAtPos(TokenType.NEWLINE, s);

                case "\\": {
                    const next = this.read();
                    switch (next) {
                        case "\\":
                        case '"':
                        case "'":
                            s += next;
                            break;

                        default:
                            return this.tokenAtPos(TokenType.BADESCAPE, c + next);
                    }

                    break;
                }

                default:
                    s += c;
            }
        }
    }

    private tokenAtPos(type: TokenType, text = ""): Token {
        return {type, pos: Object.assign({}, this.pos), text};
    }

    private unread(): void {
        if (this.last !== null && this.lastCh !== null) {
            this.text = this.lastCh + this.text;
            this.pos = this.last;
            this.last = null;
            this.lastCh = null;
        }
    }
}
